server/app/filters/CiviFormProfileFilter.java

```diff
@@ -0,0 +1,57 @@
+package filters;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static controllers.CallbackController.REDIRECT_TO_SESSION_KEY;
+import static play.mvc.Results.redirect;
+
+import akka.stream.Materializer;
+import auth.GuestClient;
+import auth.ProfileUtils;
+import com.google.inject.Inject;
+import controllers.routes;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import play.mvc.Filter;
+import play.mvc.Http;
+import play.mvc.Result;
+
+/** Filter that ensures all sessions have a CiviFormProfile. */
+public final class CiviFormProfileFilter extends Filter {
+  private final ProfileUtils profileUtils;
+
+  @Inject
+  public CiviFormProfileFilter(Materializer mat, ProfileUtils profileUtils) {
+    super(mat);
+    this.profileUtils = checkNotNull(profileUtils);
+  }
+
+  private boolean profileIsMissing(Http.RequestHeader requestHeader) {
+    return profileUtils.currentUserProfile(requestHeader).isEmpty();
+  }
+
+  private boolean shouldApplyThisFilter(Http.RequestHeader requestHeader) {
```

Comment:
(blocking) There's a number of bool checks here. Can you add some javadoc to summarize the rules?

Reply:
Done